### PR TITLE
Adding system flag Enable Experimental Features

### PR DESCRIPTION
**Pull request by soubhik-c**
## Changes proposed in this pull request

Adding system flag Enable Experimental Features
    
    Under this flag experimental features should be released. Presently, its
    DistributedIndex which by default is disabled. It is right now a system
    property and sql conf property.
    
    precheckin: TPCHDUnitTest is failing consistently apart from
    CacheBa*DUnitTest and :runQuickstart. Rest everything is clean including
    a similar failure as :runQuickstart in AQPDUnit testQuickStart.

## Patch testing

CreateIndexTest now uses a mix of system property and sqlconf property.

## ReleaseNotes.txt changes

(Does this change require an entry in ReleaseNotes.txt? If yes, has it been added to it?)

## Other PRs 

(Does this change require changes in other projects- store, spark, spark-jobserver, aqp? Add the links of PR of the other subprojects that are related to this change)

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/sources/SnappyOptimizations.scala [hunk 1]
```diff
@@ -408,7 +409,13 @@ case class ResolveIndex(implicit val snappySession: SnappySession) extends Rule[
 
   override def apply(plan: LogicalPlan): LogicalPlan = {
     val joinOrderHints = JoinOrderStrategy.getJoinOrderHints
-
+    val enabled: Boolean = io.snappydata.Property.EnableExperimentalFeatures.
+        configEntry.getConf[Boolean](snappySession.snappyContext.conf) ||
+        java.lang.Boolean.getBoolean(io.snappydata.Property.EnableExperimentalFeatures
+            .configEntry.key)
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/sources/SnappyOptimizations.scala]: I think we should avoid this explicit system property otherwise it will become a general practise which is not a good idea. I think better option will be to allow alternate "spark." (like done in Property.ValueAlt but not there for SQLVal) and let SparkConf handle those properties for both embedded and split modes. Let me see adding alternative values like done in ValueAlt for SQLVal too (or if you can do it then will be good since I have lots of other stuff to complete urgently).

File: core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala [hunk 2]
```diff
@@ -105,29 +110,30 @@ class CreateIndexTest extends SnappyFunSuite with BeforeAndAfterEach {
...
-      validateIndex(Seq(indexThree))(_)
+      CreateIndexTest.validateIndex(Seq(indexThree))(_)
     }
   }
 
   test("Test create Index on Column Table using Snappy API") {
     val tableName: String = "tcol1"
     val snContext = SnappyContext(sc)
     context.set(snContext)
+    System.setProperty(io.snappydata.Property.EnableExperimentalFeatures.configEntry.key, "true")
```
Comment (sumwale) [inline core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala]: Is this to test setting as a system property?

Comment (sumwale) [inline core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala]: The afterEach is also setting the conf property, and System property is not being cleared at the end of each test either. So if the intention was to test either of conf or System.property then this is not achieving it. Set either of the two ways in the test, then clear both at the end of each test.

**Comment by sumwale** [comment]
@soubhik-c I have the changes to SQLVal ready. Will be checked in a couple of hours.

File: core/src/main/scala/org/apache/spark/sql/sources/SnappyOptimizations.scala (hunk 1 above)
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/sql/sources/SnappyOptimizations.scala]: Ok, I pushed changes in 12119c1. So you can create this with alternative spark suffix. Then use:

```scala
EnableExperimentalFeatures.get[Boolean](snappySession.snappyContext.conf)
```

Spark prefixed system properties (and also "snappydata." ones in embedded mode) will automatically be picked up by SparkConf and from there will be copied to SQLConf by default so separate system property handling is not required and will work in split mode too.

Not required specifically for this I suppose since this is a temporary property, but this can be the general way for snappydata properties.

**Comment by sumwale** [comment]
See the last comment -- please merge master and update.

File: core/src/main/scala/io/snappydata/Literals.scala [hunk 3]
```diff
@@ -194,7 +194,13 @@ object Property extends Enumeration {
 
   val HashJoinSize: SQLType = SQLVal[Long](s"${Constant.PROPERTY_PREFIX}hashJoinSize",
     "The join would be converted into a hash join if the table is of size less" +
-        "than hashJoinSize. Default value is 100 MB.", Some(100L * 1024 * 1024))
+        "than localhashjoinsize. Default value is 100 MB.", Some(100*1024*1024))
```
Comment (sumwale) [inline core/src/main/scala/io/snappydata/Literals.scala]: revert this (probably while resolving the conflict)

File: core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala [hunk 4]
```diff
@@ -49,13 +50,16 @@ class CreateIndexTest extends SnappyFunSuite with BeforeAndAfterEach {
...
+    System.clearProperty(io.snappydata.Property.EnableExperimentalFeatures.configEntry.key)
     super.afterAll()
   }
 
   override def afterEach(): Unit = {
     try {
       val snContext = context.getAndSet(null)
       if (snContext != null) {
+        snContext.setConf(io.snappydata.Property.EnableExperimentalFeatures.configEntry.key,
+          io.snappydata.Property.EnableExperimentalFeatures.configEntry.defaultValueString)
```
Comment (sumwale) [inline core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala]: Just this should work, no? `snContext.setConf(io.snappydata.Property.EnableExperimentalFeatures.name, "false")`
I don't think explicit passing of defaultValueString is required.

File: core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala [hunk 5]
```diff
@@ -69,6 +73,7 @@ class CreateIndexTest extends SnappyFunSuite with BeforeAndAfterEach {
   test("Test choice of index according to where predicate") {
     val tableName = "tabOne"
     val snContext = SnappyContext(sc)
+    snContext.setConf(io.snappydata.Property.EnableExperimentalFeatures.configEntry.key, "true")
```
Comment (sumwale) [inline core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala]: .name instead of .configEntry.key will be good. Lets use that so it becomes a general practise for future code.

File: core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala [hunk 6]
```diff
@@ -229,6 +235,7 @@ class CreateIndexTest extends SnappyFunSuite with BeforeAndAfterEach {
   }
 
   test("Test two table joins") {
+    System.setProperty(io.snappydata.Property.EnableExperimentalFeatures.configEntry.key, "true")
```
Comment (sumwale) [inline core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala]: .name here too

File: core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala [hunk 7]
```diff
@@ -340,6 +347,7 @@ class CreateIndexTest extends SnappyFunSuite with BeforeAndAfterEach {
   }
 
   test("Test two table joins corner cases") {
+    System.setProperty(io.snappydata.Property.EnableExperimentalFeatures.configEntry.key, "true")
```
Comment (sumwale) [inline core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala]: .name here

File: core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala [hunk 8]
```diff
@@ -401,6 +409,7 @@ class CreateIndexTest extends SnappyFunSuite with BeforeAndAfterEach {
 
     val snContext = SnappyContext(sc)
     context.set(snContext)
+    snContext.setConf(io.snappydata.Property.EnableExperimentalFeatures.configEntry.key, "true")
```
Comment (sumwale) [inline core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala]: .name here

File: core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala [hunk 9]
```diff
@@ -591,6 +600,7 @@ class CreateIndexTest extends SnappyFunSuite with BeforeAndAfterEach {
...
```
Comment (sumwale) [inline core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala]: .name here

File: core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala [hunk 10]
```diff
@@ -52,10 +53,22 @@ class CreateIndexTest extends SnappyFunSuite with BeforeAndAfterEach {
     super.afterAll()
   }
 
+  override def beforeEach(): Unit = {
+    try {
+      val snContext = SnappyContext(sc)
+      snContext.setConf(io.snappydata.Property.EnableExperimentalFeatures.configEntry.key, "true")
```
Comment (sumwale) [inline core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala]: is just ".name" not working?

File: core/src/main/scala/io/snappydata/Literals.scala (hunk 3 above)
Comment (soubhik-c) [inline core/src/main/scala/io/snappydata/Literals.scala]: ah!.. yes. doing it.

File: core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala (hunk 10 above)
Comment (soubhik-c) [inline core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala]: noticed just now. doing it.
